src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -343,7 +409,173 @@ public void Stop()
             }
         }
 
+        /* W3C support functionality (see https://w3c.github.io/trace-context) */
+
+        /// <summary>
+        /// Holds the W3C 'tracestate' header as a string.   
+        /// 
+        /// Tracestate is intended to carry information supplemental to trace identity contained 
+        /// in traceparent. List of key value pairs carried by tracestate convey information 
+        /// about request position in multiple distributed tracing graphs. It is typically used 
+        /// by distributed tracing systems and should not be used as a general purpose baggage
+        /// as this use may break correlation of a distributed trace.
+        /// 
+        /// Logically it is just a kind of baggage (if flows just like baggage), but because
+        /// it is expected to be special cased (it has its own HTTP header), it is more 
+        /// convenient/efficient if it is not lumped in with other baggage.   
+        /// </summary>
+        public string TraceStateString
+        {
+            get
+            {
+                for (var activity = this; activity != null; activity = activity.Parent)
+                {
+                    var val = activity._traceState;
+                    if (val != null)
+                        return val;
+                }
+                return null;
+            }
+            set
+            {
+                _traceState = value;
+            }
+        }
+
+        /// <summary>
+        /// If the Activity has the W3C format, this returns the ID for the SPAN part of the Id.  
+        /// Otherwise it returns a zero SpanId. 
+        /// </summary>
+        public ref readonly ActivitySpanId SpanId
```

Comment:
Interesting... why ref readonly?

Reply:
This is because SpanId caches things (basically its conversion to a Hex string), and lazily populates it.   Thus to get the benefit of that caching they need to be passed by reference (It is also 1 pointer rather than sizeof(Span) which is twice as big).   It is definately true that there is performance subtelty here (it would be easy to copy it out rather than keep it as a ref), but we really only expect a modest number of clients who really need to get this right.

I have put a comment in both the SpanId and TraceId doc-comment about this to make it more explicit.

Reply:
> It is definately true that there is performance subtelty here (it would be easy to copy it out rather than keep it as a ref)

In particular, won't any method calls / property accessors accessed off of the ref readonly instance result in C# making a copy on which to make the call?